auth.go

```diff
@@ -459,3 +460,43 @@ func (h *LogoutHandler) redirect(w http.ResponseWriter, r *http.Request, idToken
 
 	http.Redirect(w, r, redirectURL.String(), http.StatusFound)
 }
+
+// Handles token refreshes.
+type RefreshHandler struct {
+	auth     *Auth
+	keychain *keychain.Keychain
+}
+
+func newRefreshHandler(auth *Auth, keychain *keychain.Keychain) http.Handler {
+	return &RefreshHandler{auth, keychain}
+}
+
+func (h *RefreshHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if !h.keychain.Guard(w, r) { // API
+		return
+	}
+
+	sessionID := r.Header.Get("Wave-Session-ID")
+	session, ok := h.auth.get(sessionID)
+
+	if !ok {
+		echo(Log{"t": "refresh_session", "error": "session unavailable"})
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
+		return
+	}
+
+	token, err := h.auth.ensureValidOAuth2Token(r.Context(), session.token)
+	if err != nil {
+		// Purge session and reload clients if refresh not successful?
+		echo(Log{"t": "refresh_session", "error": err.Error()})
+		http.Error(w, http.StatusText(http.StatusUnauthorized), http.StatusUnauthorized)
```

Comment:
Is this always unauthorized?

Reply:
I believe most of the time, but the AP server can also be down etc. Other options could be:
* propagating the err - probably not ideal security-wise.
* returning generic 500 server error.

Reply:
I vote for default on 500 unless there is a way to distinguish the cause.